extensions/kafka-eight-simpleConsumer/src/main/java/io/druid/firehose/kafka/KafkaEightSimpleConsumerFirehoseFactory.java

```diff
@@ -0,0 +1,305 @@
+      @JsonProperty("partitionIdList") List<Integer> partitionIdList,
+      @JsonProperty("clientId") String clientId,
+      @JsonProperty("feed") String feed,
+      @JsonProperty("queueBufferLength") Integer queueBufferLength,
+      @JsonProperty("resetOffsetToEarliest") Boolean resetOffsetToEarliest
+  )
+  {
+    this.brokerList = brokerList;
+    this.partitionIdList = partitionIdList;
+    this.clientId = clientId;
+    this.feed = feed;
+
+    this.queueBufferLength = queueBufferLength == null ? DEFAULT_QUEUE_BUFFER_LENGTH : queueBufferLength;
+    log.info("queueBufferLength loaded as[%s]", this.queueBufferLength);
+
+    this.earliest = resetOffsetToEarliest == null ? true : resetOffsetToEarliest.booleanValue();
+    log.info("Default behavior of cosumer set to earliest? [%s]", this.earliest);
+  }
+
+  private Map<Integer, Long> loadOffsetFromPreviousMetaData(Object lastCommit)
+  {
+    Map<Integer, Long> offsetMap = Maps.newHashMap();
+    if (lastCommit == null) {
+      return offsetMap;
+    }
+    if (lastCommit instanceof Map) {
+      Map<Object, Object> lastCommitMap = (Map) lastCommit;
+      for (Map.Entry<Object, Object> entry : lastCommitMap.entrySet()) {
+        try {
+          int partitionId = Integer.parseInt(entry.getKey().toString());
+          long offset = Long.parseLong(entry.getValue().toString());
+          log.debug("Recover last commit information partitionId [%s], offset [%s]", partitionId, offset);
+          offsetMap.put(partitionId, offset);
+        }
+        catch (NumberFormatException e) {
+          log.error(e, "Fail to load offset from previous meta data [%s]", entry);
+        }
+      }
+      log.info("Loaded offset map[%s]", offsetMap);
+    } else {
+      log.error("Unable to cast lastCommit to Map");
+    }
+    return offsetMap;
+  }
+
+  @Override
+  public FirehoseV2 connect(final ByteBufferInputRowParser firehoseParser, Object lastCommit) throws IOException
+  {
+    final Map<Integer, Long> lastOffsets = loadOffsetFromPreviousMetaData(lastCommit);
+
+    for (Integer partition : partitionIdList) {
+      final KafkaSimpleConsumer kafkaSimpleConsumer = new KafkaSimpleConsumer(
+          feed, partition, clientId, brokerList, earliest
+      );
+      Long startOffset = lastOffsets.get(partition);
+      PartitionConsumerWorker worker = new PartitionConsumerWorker(
+          feed, kafkaSimpleConsumer, partition, startOffset == null ? 0 : startOffset
+      );
+      consumerWorkers.add(worker);
+    }
+
+    final LinkedBlockingQueue<BytesMessageWithOffset> messageQueue = new LinkedBlockingQueue<BytesMessageWithOffset>(
+        queueBufferLength
+    );
+    log.info("Kicking off all consumers");
+    for (PartitionConsumerWorker worker : consumerWorkers) {
+      worker.go(messageQueue);
+    }
+    log.info("All consumer started");
+
+    return new FirehoseV2()
+    {
+      private ConcurrentMap<Integer, Long> lastOffsetPartitions;
```

Comment:
I was wrong about the interfaces. The FirehoseV2 interface actually requires that start(), advance(), currRow(), and makeCommitter() are called on the same thread. Since that is also what currently happens, I think everything is cool and there's no reason for this map to be thread safe.